pkg/operator/deploymentcontroller/deployment_controller.go

```diff
@@ -0,0 +1,199 @@
+package deploymentcontroller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	appsinformersv1 "k8s.io/client-go/informers/apps/v1"
+	"k8s.io/client-go/kubernetes"
+
+	opv1 "github.com/openshift/api/operator/v1"
+	configinformers "github.com/openshift/client-go/config/informers/externalversions"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+// DeploymentHookFunc is a hook function to modify the Deployment.
+type DeploymentHookFunc func(*opv1.OperatorSpec, *appsv1.Deployment) error
+
+// ManifestHookFunc is a hook function to modify the manifest in raw format.
+type ManifestHookFunc func(*opv1.OperatorSpec, []byte) ([]byte, error)
+
+// DeploymentController is a generic controller that manages a deployment
+
+//
+// <name>Available: indicates that the deployment controller  was successfully deployed and at least one Deployment replica is available.
+// <name>Progressing: indicates that the CSI Controller Service is being deployed.
+// <name>Degraded: produced when the sync() method returns an error.
+type DeploymentController struct {
+	name           string
+	manifest       []byte
+	operatorClient v1helpers.OperatorClient
+	kubeClient     kubernetes.Interface
+	deployInformer appsinformersv1.DeploymentInformer
+	configInformer configinformers.SharedInformerFactory
+	// Optional hook functions to modify the deployment manifest.
+	// If one of these functions returns an error, the sync
+	// fails indicating the ordinal position of the failed function.
+	// Also, in that scenario the Degraded status is set to True.
+	// TODO: Collapse this into optional deployment hook.
+	optionalManifestHooks []ManifestHookFunc
+	// Optional hook functions to modify the Deployment.
+	// If one of these functions returns an error, the sync
+	// fails indicating the ordinal position of the failed function.
+	// Also, in that scenario the Degraded status is set to True.
+	optionalDeploymentHooks []DeploymentHookFunc
+}
+
+func New(
+	name string,
+	manifest []byte,
+	recorder events.Recorder,
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	deployInformer appsinformersv1.DeploymentInformer,
+	optionalInformers []factory.Informer,
+	optionalManifestHooks []ManifestHookFunc,
```

Comment:
doc the difference between this and the deploymenthooks.  The construction here is good and I understand why, but be sure to explain it godoc for others.

Reply:
Sure, I explained it [here](https://github.com/openshift/library-go/pull/1095/files#diff-abb679a52a6ea86c466b84d87793ee37682c0e4b6a111234e69cb684d7ec3157R44)